### ISSUE TITLE
Feature Request: Delayed Messages

### ISSUE THREAD

**Issue by firrae** [nats-io/gnatsd#514]
- [ ] Defect
 - [x] Feature Request or Change Proposal
 
#### Use Case: 
A scheduled event that you want to put in the queue and have delivered on a specified delay from pub time.

Specifically, my use case is as follows: We have a series of events that users can trigger that have an option for a delay. We currently put the events into a queue with a trigger time and the workers for those events check the time and if it's not at or past that time yet it dumps it back in the queue. This causes an issue where events sometimes get stuck behind other longer events and eat up workers for the amount of time it takes to pull the queue item, check, and put it back. It also opens the potential for a loop that artificially inflates our usage statistics (this is something we could handle except:) and we can no longer effectively auto-scale our worker assignments.
 
#### Proposed Change:
I propose an option that is handled by the server where on message creation an additional option is passed by clients, defaulting to 0, that outlines a delay in milliseconds for how long to delay the message from creation to first attempted delivery.
 
#### Who Benefits From The Change(s)?
I believe a number of people could see a benefit from this, namely, people and groups that are developing distributed processing systems, and communication systems (email, SMS, phone, etc.).
 
#### Alternative Approaches
Like I mentioned in my use case above the application/worker could currently handle this just not as nicely as having it inside the core of the server. In my current use, I'm checking a time in the message structure on the worker and putting it back in the queue if the time hasn't passed yet. Alternatively, if your language of choice has the good time management, you could build a process into your application that internally checks if it's time is met and sends it to the queue then, though this seems to be the least efficient method I've come up with so far.

Thanks for considering,

Steve

**Comment by tylertreat** [comment]
I think this request would be more appropriate for [NATS Streaming](https://github.com/nats-io/nats-streaming-server). Reason being NATS by design does not store messages and avoids holding them in memory (beyond just network buffering) in order to maintain its high performance. If NATS had to store messages and do bookkeeping for scheduling delivery, I suspect its performance would be affected. Moreover, since NATS is ephemeral, it would be effectively the same to just have the client be responsible for delayed delivery. That is, if you publish a delayed message to NATS and NATS schedules its delivery but goes down before delivering it (autoscale event, hardware failure, etc.), handling delivery at the client would actually be *more* resilient. Similarly, if you publish a delayed message to NATS and NATS delivers it in the future, there's no guarantee a subscriber actually receives the message because, again, NATS is at-most-once delivery.

NATS Streaming, on the other hand, stores every message. Thus, it provides a higher QoS with at-least-once delivery. It already has to do delivery bookkeeping, and it ensures subscribers actually receive messages. As a result, I think delayed delivery makes more sense in the context of NATS Streaming.

For what it's worth, Amazon SQS provides a similar feature called [Delay Queues](http://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-delay-queues.html), though this is limited to a static delay per queue rather than per message. I suspect this would be much easier to implement in NATS Streaming (or rather impact performance the least).

**Comment by firrae** [comment]
Hm, fair enough. I agree there isn't much out there currently but I have found a number of posts and tutorials on how to set up such a thing in RabbitMQ.

To be more specific SQS does do this idea via [message timers](http://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-message-timers.html) when you create individual messages. That is maybe the best solution I have found so far, but sadly SQS isn't a viable option as we deploy on closed networks so there's no internet access.

I will take this over to NATS Streaming and see what the discussion is like there tho as you suggested as after reading it I agree that it is likely the better candidate for such a feature to be implemented, I just hadn't read much into it.
